README.rst
django-intercooler_helpers
==========================

:author: Keryn Knight
:version: 0.2.0

.. |travis_stable| image:: https://travis-ci.org/kezabelle/django-intercoolerjs-helpers.svg?branch=0.2.0
  :target: https://travis-ci.org/kezabelle/django-intercoolerjs-helpers

.. |travis_master| image:: https://travis-ci.org/kezabelle/django-intercoolerjs-helpers.svg?branch=master
  :target: https://travis-ci.org/kezabelle/django-intercoolerjs-helpers

==============  ======
Release         Status
==============  ======
stable (0.2.0)  |travis_stable|
master          |travis_master|
==============  ======


.. contents:: Sections
   :depth: 2

What it does
------------

``intercooler_helpers`` is a small reusable app for `Django`_ which provides a
few improvements for working with `Intercooler.js`_.

It providea a middleware which extracts relevant `Intercooler.js`_ data from the
querystring, and attaches it to the request as a separate ``QueryDict`` (ie: it
behaves like ``request.POST`` or ``request.GET``)

It also provides a small middleware for changing request method based on either the
query string (``_method=PUT``) or a request header(``X-HTTP-Method-Override: PUT``)

Between them, they should capture all the incoming `Intercooler.js`_ data on
which the server may act.

Installation and usage
----------------------

This application depends on `django-intercoolerjs`_ which provides a copy of
`Intercooler.js`_ bundled up for use with the standard `Django`_ staticfiles
application.

Installation
^^^^^^^^^^^^

You can grab ``0.2.0`` from `PyPI`_ like so::

  pip install django-intercooler-helpers==0.2.0

Or you can grab it from  `GitHub`_  like this::

  pip install -e git+https://github.com/kezabelle/django-intercooler-helpers.git#egg=django-intercooler-helpers

Configuration
^^^^^^^^^^^^^
You need to add ``intercooler_helpers.middleware.IntercoolerData`` to your
``MIDDLEWARE_CLASSES`` (or ``MIDDLEWARE`` on Django **1.10+**).

You may optionally want to add ``intercooler_helpers.middleware.HttpMethodOverride``
as well, if you don't already have a method by which to fake the HTTP Method change.
If you're using ``<meta name="intercoolerjs:use-actual-http-method" content="true"/>``
then you don't need ``HttpMethodOverride`` at all.

To install all the middleware, you want something like::

  # MIDDLEWARE = ... for newer Django
  MIDDLEWARE_CLASSES = (
    # ...
    'intercooler_helpers.middleware.HttpMethodOverride',
    'intercooler_helpers.middleware.IntercoolerData',
    'intercooler_helpers.middleware.IntercoolerRedirector',
    # ...
  )

``HttpMethodOverride`` and ``IntercoolerData`` ought to be near the top of the iterable, as they both make use of ``process_request(request)``.
``IntercoolerRedirector`` ought to be near the bottom, as it operates on ``process_response(request, response)`` and you probably want to convert the response to a client-side redirect at the earliest opportunity.

Usage
^^^^^

A brief overview of the public API provided so far:

IntercoolerData
***************

For fully correct detection of `Intercooler.js`_ requests, you can call
``request.is_intercooler()``.
Behind the scenes, it uses ``request.maybe_intercooler()`` to
detect whether ``ic-request`` was present, indicating it *may* have been a
valid `Intercooler.js`_ request, and also checks ``request.is_ajax()``

To parse the Intercooler-related data out of the query-string, you can use
``request.intercooler_data`` (not a method!) which is a ``QueryDict`` and should
behave exactly like ``request.GET`` - It pulls all of the ``ic-*`` keys out
of ``request.GET`` and puts them in a separate data structure, leaving
your ``request.GET`` cleaned of extraenous data.

``request.intercooler_data`` is a **lazy** data structure, like ``request.user``,
so will not modify ``request.GET`` until access is attempted.

The following properties exist, mapping back to the keys mentioned in the
`Intercooler.js Reference document`_

- ``request.intercooler_data.url`` returns a ``namedtuple`` containing

  - returns the ``ic-current-url`` (converted via ``urlparse``) or ``None``
  - A `Django`_ ``ResolverMatch`` pointing to the view which made the request (based on ``ic-current-url``) or ``None``
- ``request.intercooler_data.element`` returns a ``namedtuple`` containing

  -  ``ic-element-name`` or ``None``
  -  ``ic-element-id`` or ``None``
- ``request.intercooler_data.id``

  - the ``ic-id`` which made the request. an ever-incrementing integer.
- ``request.intercooler_data.request``

  - a boolean indicating that it was an `Intercooler.js`_ request. Should always
    be true if ``request.is_intercooler()`` said so.
- ``request.intercooler_data.target_id``

  -  ``ic-target-id`` or ``None``
- ``request.intercooler_data.trigger`` returns a ``namedtuple`` containing

  -  ``ic-trigger-name`` or ``None``
  -  ``ic-trigger-id`` or ``None``
- ``request.intercooler_data.prompt_value``

  - If no ``ic-prompt-name`` was given and a prompt was used, this will contain
    the user's response. Appears to be undocumented?


HttpMethodOverride
******************

- ``request.changed_method`` is a boolean indicating that the request was
  toggled from being a ``POST`` to something else (one of
  ``GET``, ``HEAD``, ``POST``, ``PUT``, ``PATCH``, ``DELETE``, ``OPTIONS`` ...
  though why you'd want to ``POST`` and have it act as a ``GET`` is beyond me.
  But that's your choice)
- ``request.original_method`` if either ``_method=X`` or
  ``X-HTTP-Method-Override: X`` caused the request to change method, then this
  will contain the original request. It should always be ``POST``
- ``request.method`` will reflect the desired HTTP method, rather than the one
  originally used (``POST``)


IntercoolerRedirector
*********************

If a redirect status code is given (> 300, < 400), and the request originated from `Intercooler.js`_ (assumes ``IntercoolerData`` is installed so that ``request.is_intercooler()`` may be called), remove the ``Location`` header from the response, and create a new ``HttpResponse`` with all the other headers, and also the ``X-IC-Redirect`` header to indicate to `Intercooler.js`_ that it needs to do a client side-redirect.


Supported Django versions
-------------------------

The tests are run against Django 1.8 through 1.10, and Python 2.7, 3.3, 3.4 and 3.5.

Running the tests
^^^^^^^^^^^^^^^^^

If you have a cloned copy, you can do::

  python setup.py test

If you have tox, you can just do::

  tox

Running the demo
^^^^^^^^^^^^^^^^

A barebones demo is provided. It assumes you're using something like `virtualenv`_ and
`virtualenvwrapper`_ but you can probably figure it out otherwise::

    mktmpenv --python=`which python3`
    pip install -e git+https://github.com/kezabelle/django-intercooler-helpers.git#egg=django-intercooler-helpers

Then probably::

    cd src/django-intercooler-helpers
    python demo_project.py runserver

It shows off a few of the same demos that the `Intercooler.js`_ website does.

Contributing
------------

Please do!

The project is hosted on `GitHub`_ in the `kezabelle/django-intercooler-helpers`_
repository.

Bug reports and feature requests can be filed on the repository's `issue tracker`_.

If something can be discussed in 140 character chunks, there's also `my Twitter account`_.

Roadmap
-------

TODO.

The license
-----------

It's `FreeBSD`_. There's should be a ``LICENSE`` file in the root of the repository, and in any archives.

.. _FreeBSD: http://en.wikipedia.org/wiki/BSD_licenses#2-clause_license_.28.22Simplified_BSD_License.22_or_.22FreeBSD_License.22.29
.. _Django: https://www.djangoproject.com/
.. _Intercooler.js: http://intercoolerjs.org/
.. _django-intercoolerjs: https://github.com/brejoc/django-intercoolerjs
.. _GitHub: https://github.com/
.. _PyPI: https://pypi.python.org/pypi
.. _Intercooler.js Reference document: http://intercoolerjs.org/reference.html
.. _virtualenvwrapper: https://virtualenvwrapper.readthedocs.io/en/latest/
.. _virtualenv: https://virtualenv.pypa.io/en/stable/
.. _kezabelle/django-intercooler-helpers: https://github.com/kezabelle/django-intercooler-helpers/
.. _issue tracker: https://github.com/kezabelle/django-intercooler-helpers/issues/
.. _my Twitter account: https://twitter.com/kezabelle/


CHANGELOG
Change history for django-intercooler_helpers
-------------------------------------------------------------
0.2.0
^^^^^^
* Initial release.


test_urls.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals, absolute_import

from uuid import uuid4

from django.conf.urls import url
from django.forms import Form, CharField, IntegerField
from django.http import HttpResponse, Http404
from django.shortcuts import redirect
from django.template.defaultfilters import pluralize
from django.template.response import TemplateResponse
try:
    from django.urls import reverse
except ImportError:  # Django <1.10
    from django.core.urlresolvers import reverse


def _page_data():
    return tuple(str(uuid4()) for x in range(0, 10))


clicktrack = 0

def click(request):
    global clicktrack
    clicktrack += 1
    do_reset = (request.is_intercooler() and
                request.intercooler_data.element.id == 'intro-btn2' and
                request.intercooler_data.current_url.match is not None)
    if do_reset:
        clicktrack = 0
    time = pluralize(clicktrack)
    text = "<span>You clicked me {} time{}...</span>".format(clicktrack, time)
    if do_reset:
        text = "<span>You reset the counter!, via {}</span>".format(request.intercooler_data.trigger.id)
    if not request.is_intercooler():
        raise Http404("Not allowed to come here outside of an Intercooler.js request!")
    resp = HttpResponse(text)
    return resp


def redirector(request):
    return redirect(reverse('redirected'))


def redirected(request):
    return TemplateResponse(request, template="redirected.html", context={})


class TestForm(Form):
    field = CharField()
    number = IntegerField(max_value=10, min_value=5)


def form(request):
    template = "form.html"
    _form = TestForm(request.POST or None)
    if _form.is_valid():
        return redirect(reverse('redirected'))
    context = {'form': _form}
    return TemplateResponse(request, template=template, context=context)



def polling_stop(request):
    resp = HttpResponse("Cancelled")
    resp['X-IC-CancelPolling'] = "true"
    return resp


def polling_start(request):
    resp = HttpResponse("")
    resp['X-IC-ResumePolling'] = "true"
    return resp


def polling(request):
    template = "polling.html"
    if request.is_intercooler():
        template = "polling_response.html"
    context = {
        'item': str(uuid4()),
    }
    return TemplateResponse(request, template=template, context=context)


def infinite_scrolling(request):
    template = "infinite_scrolling.html"
    if request.is_intercooler():
        template = "infinite_scrolling_include.html"
    context = {
        'rows': _page_data(),
    }
    return TemplateResponse(request, template=template, context=context)


def root(request):
    template = "demo_project.html"
    context = {
        'rows': _page_data(),
        'form': TestForm()
    }
    return TemplateResponse(request, template=template, context=context)


urlpatterns = [
    url('^form/$', form, name='form'),
    url('^redirector/redirected/$', redirected, name='redirected'),
    url('^redirector/$', redirector, name='redirector'),
    url('^click/$', click, name='click'),
    url('^polling/stop/$', polling_stop, name='polling_stop'),
    url('^polling/start/$', polling_start, name='polling_start'),
    url('^polling/$', polling, name='polling'),
    url('^infinite/scrolling/$', infinite_scrolling, name='infinite_scrolling'),
    url('^$', root, name='root'),
]



test_settings.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals, absolute_import
import os

DEBUG = os.environ.get('DEBUG', 'on') == 'on'
SECRET_KEY = os.environ.get('SECRET_KEY', 'TESTTESTTESTTESTTESTTESTTESTTEST')
ALLOWED_HOSTS = os.environ.get('ALLOWED_HOSTS', 'localhost,testserver,*').split(',')
BASE_DIR = os.path.abspath(os.path.dirname(os.path.abspath(__file__)))

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': os.path.join(BASE_DIR, 'db.sqlite3'),
    }
}

INSTALLED_APPS = [
    'django.contrib.sessions',
    'django.contrib.contenttypes',
    'django.contrib.admin',
    'django.contrib.staticfiles',
    'django.contrib.auth',
    'intercoolerjs',
    'intercooler_helpers',
]

SKIP_SOUTH_TESTS = True
SOUTH_TESTS_MIGRATE = False

STATIC_URL = '/__static__/'
MEDIA_URL = '/__media__/'
MESSAGE_STORAGE = 'django.contrib.messages.storage.cookie.CookieStorage'
SESSION_ENGINE = 'django.contrib.sessions.backends.signed_cookies'
SESSION_COOKIE_HTTPONLY = True

ROOT_URLCONF = 'test_urls'

# Use a fast hasher to speed up tests.
PASSWORD_HASHERS = (
    'django.contrib.auth.hashers.MD5PasswordHasher',
)

SITE_ID = 1

TEMPLATE_CONTEXT_PROCESSORS = (
    'django.contrib.auth.context_processors.auth',
)
TEMPLATE_LOADERS = (
    'django.template.loaders.filesystem.Loader',
)
TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [os.path.join(BASE_DIR, "test_templates")],
        'OPTIONS': {
            'context_processors': TEMPLATE_CONTEXT_PROCESSORS,
            'loaders': TEMPLATE_LOADERS,
        },
    },
]

MIDDLEWARE_CLASSES = (
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'intercooler_helpers.middleware.HttpMethodOverride',
    'intercooler_helpers.middleware.IntercoolerData',
    'intercooler_helpers.middleware.IntercoolerRedirector',
    'intercooler_helpers.middleware.SelectFromResponse',
)

STATIC_ROOT = os.path.join(BASE_DIR, 'test_collectstatic')
MEDIA_ROOT = os.path.join(BASE_DIR, 'test_media')

USE_TZ = True

SILENCED_SYSTEM_CHECKS = ['1_8.W001']


intercooler_helpers/tests/__init__.py
# -*- coding: utf-8 -*-
from __future__ import absolute_import, unicode_literals
__all__ = []


demo_project.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
from __future__ import absolute_import
import os
import sys
sys.dont_write_bytecode = True
MISSING_DEPENDENCIES = []
try:
    from django.conf import settings
except ImportError:
    MISSING_DEPENDENCIES.append("Django")
try:
    import intercoolerjs
except ImportError:
    MISSING_DEPENDENCIES.append("django-intercoolerjs")

if MISSING_DEPENDENCIES:
    deps = " ".join(MISSING_DEPENDENCIES)
    sys.stdout.write("You'll need to `pip install {}` to run this demo\n".format(deps))
    sys.exit(1)

os.environ.setdefault("DJANGO_SETTINGS_MODULE", "test_settings")

from django.core.wsgi import get_wsgi_application
application = get_wsgi_application()

if __name__ == "__main__":
    from django.core.management import execute_from_command_line
    execute_from_command_line(sys.argv)



intercooler_helpers/middleware.py
# -*- coding: utf-8 -*-
from __future__ import absolute_import, unicode_literals

from collections import namedtuple
from contextlib import contextmanager

from django.http import QueryDict, HttpResponse
try:
    from django.urls import Resolver404, resolve
except ImportError:  # Django <1.10
    from django.core.urlresolvers import Resolver404, resolve
from django.utils.functional import SimpleLazyObject
from django.utils.six.moves.urllib.parse import urlparse
try:
    from django.utils.deprecation import MiddlewareMixin
except ImportError:  # < Django 1.10
    class MiddlewareMixin(object):
        pass


__all__ = ['IntercoolerData', 'HttpMethodOverride']


class HttpMethodOverride(MiddlewareMixin):
    """
    Support for X-HTTP-Method-Override and _method=PUT style request method
    changing.

    Note: if https://pypi.python.org/pypi/django-method-override gets updated
    with support for newer Django (ie: implements MiddlewareMixin), without
    dropping older versions, I could possibly replace this with that.
    """
    def process_request(self, request):
        request.changed_method = False
        if request.method != 'POST':
            return
        methods = {'GET', 'HEAD', 'POST', 'PUT', 'PATCH', 'DELETE', 'OPTIONS'}
        potentials = ((request.META, 'HTTP_X_HTTP_METHOD_OVERRIDE'),
                      (request.GET, '_method'),
                      (request.POST, '_method'))
        for querydict, key in potentials:
            if key in querydict and querydict[key].upper() in methods:
                newmethod = querydict[key].upper()
                # Don't change the method data if the calling method was
                # the same as the indended method.
                if newmethod == request.method:
                    return
                request.original_method = request.method
                if hasattr(querydict, '_mutable'):
                    with _mutate_querydict(querydict):
                        querydict.pop(key)
                if not hasattr(request, newmethod):
                    setattr(request, newmethod, request.POST)
                request.method = newmethod
                request.changed_method = True
                return


def _maybe_intercooler(self):
    return self.META.get('HTTP_X_IC_REQUEST') == 'true'


def _is_intercooler(self):
    return self.is_ajax() and self.maybe_intercooler()


@contextmanager
def _mutate_querydict(qd):
    qd._mutable = True
    yield qd
    qd._mutable = False


NameId = namedtuple('NameId', 'name id')
UrlMatch = namedtuple('UrlMatch', 'url match')


class IntercoolerQueryDict(QueryDict):
    @property
    def url(self):
        url = self.get('ic-current-url', None)
        match = None
        if url is not None:
            url = url.strip()
            url = urlparse(url)
            if url.path:
                try:
                    match = resolve(url.path)
                except Resolver404:
                    pass
        return UrlMatch(url, match)

    current_url = url

    @property
    def element(self):
        return NameId(self.get('ic-element-name', None), self.get('ic-element-id', None))

    @property
    def id(self):
        # I know IC calls it a UUID internally, buts its just 1, incrementing.
        return int(self.get('ic-id', '0'))

    @property
    def request(self):
        return bool(self.get('ic-request', None))

    @property
    def target_id(self):
        return self.get('ic-target-id', None)

    @property
    def trigger(self):
        return NameId(self.get('ic-trigger-name', None), self.get('ic-trigger-id', None))

    @property
    def prompt_value(self):
        return self.get('ic-prompt-value', None)

    def __repr__(self):
        props = ('id', 'request', 'target_id', 'element', 'trigger',
                 'prompt_value', 'url')
        attrs = ['{name!s}={val!r}'.format(name=prop, val=getattr(self, prop))
                 for prop in props]
        return "<{cls!s}: {attrs!s}>".format(cls=self.__class__.__name__,
                                             attrs=", ".join(attrs))


def intercooler_data(self):
    if not hasattr(self, '_processed_intercooler_data'):
        IC_KEYS = ['ic-current-url', 'ic-element-id', 'ic-element-name',
                   'ic-id', 'ic-prompt-value', 'ic-target-id',
                   'ic-trigger-id', 'ic-trigger-name', 'ic-request']
        ic_qd = IntercoolerQueryDict('', encoding=self.encoding)
        if self.method in ('GET', 'HEAD', 'OPTIONS'):
            query_params = self.GET
        else:
            query_params = self.POST
        query_keys = tuple(query_params.keys())
        for ic_key in IC_KEYS:
            if ic_key in query_keys:
                # emulate how .get() behaves, because pop returns the
                # whole shebang.
                # For a little while, we need to pop data out of request.GET
                with _mutate_querydict(query_params) as REQUEST_DATA:
                    try:
                        removed = REQUEST_DATA.pop(ic_key)[-1]
                    except IndexError:
                        removed = []
                with _mutate_querydict(ic_qd) as IC_DATA:
                    IC_DATA.update({ic_key: removed})
        # Don't pop these ones off, so that decisions can be made for
        # handling _method
        ic_request = query_params.get('_method')
        with _mutate_querydict(ic_qd) as IC_DATA:
            IC_DATA.update({'_method': ic_request})
        # If HttpMethodOverride is in the middleware stack, this may
        # return True.
        IC_DATA.changed_method = getattr(self, 'changed_method', False)
        self._processed_intercooler_data = ic_qd
    return self._processed_intercooler_data


class IntercoolerData(MiddlewareMixin):
    def process_request(self, request):
        request.maybe_intercooler = _maybe_intercooler.__get__(request)
        request.is_intercooler = _is_intercooler.__get__(request)
        request.intercooler_data = SimpleLazyObject(intercooler_data.__get__(request))



class IntercoolerRedirector(MiddlewareMixin):
    def process_response(self, request, response):
        if not request.is_intercooler():
            return response
        if response.status_code > 300 and response.status_code < 400:
            if response.has_header('Location'):
                url = response['Location']
                del response['Location']
                new_resp = HttpResponse()
                for k, v in response.items():
                    new_resp[k] = v
                new_resp['X-IC-Redirect'] = url
                return new_resp
        return response


class SelectFromResponse(MiddlewareMixin):
    def process_response(self, request, response):
        if not request.is_intercooler():
            return response
